src/common/snippets/src/pass/transpose_decomposition.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <snippets/pass/transpose_decomposition.hpp>
+#include <snippets/itt.hpp>
+#include <snippets/snippets_isa.hpp>
+#include <snippets/pass/loop_helpers.hpp>
+
+#include <ngraph/opsets/opset1.hpp>
+#include <ngraph/opsets/opset5.hpp>
+#include <ngraph/partial_shape.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pass/manager.hpp>
+#include <numeric>
+const std::set<std::vector<int>> ngraph::snippets::pass::TransposeDecomposition::supported_cases = {{0, 2, 3, 1}};
+ngraph::snippets::pass::TransposeDecomposition::TransposeDecomposition() {
+    MATCHER_SCOPE(TransposeDecomposition);
+    // todo: we need a special transformation that detects and propagates data access pattern to Parameters and Results
+    //  this is needed to communicate access pattern to the plugin node and op::Kernel
+    // This is the reason we me match only to Parameter, this limitation could be relaxed if we propagate access pattern
+    // to the appropriate parameter
+    auto match_data = ngraph::pattern::any_input();
+    auto match_order = ngraph::pattern::wrap_type<opset1::Constant>();
+    auto match_transpose = ngraph::pattern::wrap_type<ngraph::opset1::Transpose>({match_data, match_order});
+
+    ngraph::matcher_pass_callback callback = [=](ngraph::pattern::Matcher& m) {
+        OV_ITT_SCOPED_TASK(ngraph::pass::itt::domains::SnippetsTransform, "Snippets::op::TransposeDecomposition")
+        auto& pattern_to_output = m.get_pattern_value_map();
+        const auto transpose = ov::as_type_ptr<ngraph::opset1::Transpose>(
+                                                            pattern_to_output.at(match_transpose).get_node_shared_ptr());
+
+        const auto order = ov::as_type_ptr<ov::op::v0::Constant>(pattern_to_output.at(match_order).get_node_shared_ptr());
+        if (transformation_callback(transpose) || transpose->is_dynamic() ||
+            order->get_output_element_type(0) != ov::element::i32)
+            return false;
+
+        auto order_value = order->get_vector<int>();
+        if (supported_cases.count(order_value) == 0)
+            throw ngraph::ngraph_error("TransposeDecomposition: unsupported order");
+
+        auto data_input = pattern_to_output.at(match_data);
+        const auto& data_node = pattern_to_output.at(match_data).get_node_shared_ptr();
+        // todo: we can decompose any transpose, but if it's not after parameter then it's not MHA-like pattern
+        //  and this Transpose is likely tokenized by error. So this exception is more like a sanity check:
+        //  if it's thrown => make sure that tokenization worked as expected
+        if (!ov::is_type<opset1::Parameter>(data_node))
+            throw ngraph_error("TransposeDecomposition: it's valid to decompose Transpose only after Parameter, not after " +
```

Comment:
As described in the comment above, we do this check to make sure that tokenization worked as expected.
If we wrap `match_data` as a Parameter, we skip potential transposes in this transformation that could be incorrectly tokenized.
What happens in this case after this transformation? If we get some exception with clear description (like "Transpose op don't find in jitters_map") I suppose we can wrap `match_data` as a Parameter. Otherwise, I would leave this check as is

Reply:
Yes, it will be smth like "Emmiter for Transpose is not found", so yes, let's just update the matcher. Only `match_data`, not `match_order`